hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/ImportOperation.java

```diff
@@ -0,0 +1,255 @@
+    return status;
+  }
+
+  @Override
+  public Long execute() throws IOException {
+    final long items;
+    if (status.isFile()) {
+      PathMetadata meta = new PathMetadata(status);
+      getStore().put(meta, null);
+      items = 1;
+    } else {
+      try (DurationInfo ignored =
+               new DurationInfo(LOG, "Importing %s", getStatus().getPath())) {
+        items = importDir();
+      }
+    }
+    return items;
+  }
+
+  /**
+   * Recursively import every path under path.
+   * @return number of items inserted into MetadataStore
+   * @throws IOException on I/O errors.
+   */
+  private long importDir() throws IOException {
+    Preconditions.checkArgument(status.isDirectory());
+    long items;
+    final Path basePath = status.getPath();
+    final MetadataStore ms = getStore();
+    try (BulkOperationState operationState = ms
+        .initiateBulkWrite(
+            BulkOperationState.OperationType.Import,
+            basePath)) {
+      RemoteIterator<S3ALocatedFileStatus> it = getFilesystem()
+          .listFilesAndEmptyDirectories(basePath, true);
+      items = 0;
+
+      while (it.hasNext()) {
+        S3ALocatedFileStatus located = it.next();
+        S3AFileStatus child;
+        final Path path = located.getPath();
+        final boolean isDirectory = located.isDirectory();
+        if (isDirectory) {
+          child = DynamoDBMetadataStore.makeDirStatus(path,
+              located.getOwner());
+          dirCache.add(path);
+        } else {
+          child = located.toS3AFileStatus();
+        }
+
+        putParentsIfNotPresent(child, operationState);
+
+        // We don't blindly overwrite any existing file entry in S3Guard with a
+        // new one, Because that may lose the version information.
+        // instead we merge them
+        if (!isDirectory) {
+          final PathMetadata existingEntry = S3Guard.getWithTtl(ms, path, null,
+              false);
+          if (existingEntry != null) {
+            final S3AFileStatus existingStatus = existingEntry.getFileStatus();
+            if (existingStatus.isFile()) {
+              // source is also a file.
+              // we only worry about an update if the timestamp is different,
+              final String existingEtag = existingStatus.getETag();
+              final String childEtag = child.getETag();
+              if (child.getModificationTime()
+                  != existingStatus.getModificationTime()
+                  || existingStatus.getLen() != child.getLen()
+                  || existingEtag == null
+                  || !existingEtag.equals(childEtag)) {
+                // files are potentially different, though a modtime change
+                // can just be a clock skew problem
+                // so if the etag is unchanged, we propagate any versionID
+                if (childEtag.equals(existingEtag)) {
+                  // copy over any version ID.
+                  child.setVersionId(existingStatus.getVersionId());
```

Comment:
if the etag is matching but the version marker is there and it's different that's an error, isn't it?
Also, you can suspend creating version ids on s3. What if the user wants to suspend this, and then do an import?

Reply:
I don't know. What if you overwrite a file with a new file of the same version. Maybe we'll just leave etags alone